### ISSUE TITLE
strange warning about mixed memory format

### ISSUE THREAD

**Issue by FateScript** [pytorch/pytorch#42300]
## To Reproduce

1. I tried code below, and I got warning like
```
torch/autograd/__init__.py:127: UserWarning: Mixed memory format inputs detected while calling the operator. 
The operator will output contiguous tensor even if some of the inputs are in channels_last format. 
(Triggered internally at  /pytorch/aten/src/ATen/native/TensorIterator.cpp:918.)
allow_unreachable=True)  # allow_unreachable flag
```
2. Then I comment the second nn.ReLU in self.bbox_subnet, rerun the code, and I got a warning free running.
This is strange for me.

Here is the code:
```Python
import torch
from torch import nn


class Head(nn.Module):
    def __init__(self):
        super().__init__()
        in_channels = 4
        self.cls_subnet = nn.Sequential(*[
            nn.Conv2d(in_channels, in_channels, kernel_size=3, padding=1),
            nn.ReLU(),
        ])
        self.bbox_subnet = nn.Sequential(*[
            nn.Conv2d(in_channels, in_channels, kernel_size=3, padding=1),
            nn.ReLU(),
        ])

    def forward(self, x):
        cls_subnet = self.cls_subnet(x)
        bbox_subnet = self.bbox_subnet(x)
        return cls_subnet, bbox_subnet


if __name__ == "__main__":
    x = torch.randn(2, 4, 100, 152)
    x.requires_grad_()
    head = Head()
    cls, box = head(x)
    N, _, H, W = box.shape
    tensor = box.reshape(N, -1, 4, H, W).permute(0, 3, 4, 1, 2)
    tensor.mean().backward()
```

## Expected behavior
A warning-free running.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0): 1.6
 - OS (e.g., Linux): Ubuntu
 - How you installed PyTorch (`conda`, `pip`, source): pip install torch==1.6.0+cu101 torchvision==0.7.0+cu101 -f https://download.pytorch.org/whl/torch_stable.html
 - Build command you used (if compiling from source): 
 - Python version: 3.6.8
 - CUDA/cuDNN version: 10.1
 - GPU models and configuration: 2080Ti
 - Any other relevant information:

## Additional context

cc @VitalyFedyunin @jamesr66a

**Comment by FateScript** [comment]
here is a better implementation to reproduce:
```Python
import torch
from torch import nn


class Head(nn.Module):
    def __init__(self):
        super().__init__()
        in_channels = 4
        self.bbox_subnet = nn.Sequential(*[
            nn.Conv2d(in_channels, in_channels, kernel_size=3, padding=1),
            nn.ReLU(),
        ])

    def forward(self, x):
        bbox_subnet = self.bbox_subnet(x)
        return bbox_subnet


if __name__ == "__main__":
    x = torch.randn(2, 4, 100, 152)
    x.requires_grad_()
    head = Head()
    box = head(x)
    N, _, H, W = box.shape
    tensor = box.reshape(N, -1, 4, H, W).permute(0, 3, 4, 1, 2)
    tensor.mean().backward()
```
It seems that only if you use elementwise operation(such as nn.Sigmoid, etc.) at last, This warning will be displayed.

**Comment by gchanan** [comment]
this looks expected.  It's not a good warning, though, because it doesn't tell you what to do about it.

**Comment by FateScript** [comment]
> this looks expected. It's not a good warning, though, because it doesn't tell you what to do about it.

I wonder why commenting elementwise operation(for example, nn.ReLU) could make this warning disappear.

**Comment by VitalyFedyunin** [comment]
It is happening because ` tensor = box.reshape(N, -1, 4, H, W).permute(0, 3, 4, 1, 2)` outputs channels last 3d tensor. And when it comes in the mix with non channels last model this warning appears. If it makes you nervous, you can add `tensor = box.reshape(N, -1, 4, H, W).permute(0, 3, 4, 1, 2).contiguous()` to make this warning disappear.

**Comment by FateScript** [comment]
> It is happening because ` tensor = box.reshape(N, -1, 4, H, W).permute(0, 3, 4, 1, 2)` outputs channels last 3d tensor. And when it comes in the mix with non channels last model this warning appears.
> 
> If this warning makes you nervous, you can add `tensor = box.reshape(N, -1, 4, H, W).permute(0, 3, 4, 1, 2).contiguous()` to make this warning disappear.

Actually I tried but this warning still exists and it seems that this warning is caused by elementwise operation.

As far as I know, commenting nn.ReLU in my example code is the only way to make this warning disappear. This shouldn't be an expected behaviour.

**Comment by xiaohu2015** [comment]
I also met the issue:

`
[W TensorIterator.cpp:918] Warning: Mixed memory format inputs detected while calling the operator. The operator will output contiguous tensor 
even if some of the inputs are in channels_last format. (function operator())
[W TensorIterator.cpp:918] Warning: Mixed memory format inputs detected while calling the operator. The operator will output contiguous tensor 
even if some of the inputs are in channels_last format. (function operator())
`

**Comment by drmeerkat** [comment]
Does anyone have a solution for this? I also met this error.

**Comment by mranzinger** [comment]
Also seeing this warning, and it seems like a potentially dangerous one. Is there any way to trap the forward operator that's causing it to trigger in backward?

It seems like this could possibly be related to this, which at least I consider problematic:
```
In [1]: import torch                                                                                                                                                                                                                                                                                                                                                                                                                     
In [2]: t = torch.empty(12, 1, 40, 40)                                                                                                                                                                                                                                                                                                                                                                                                   
In [3]: t.is_contiguous()                                                                                                                                                                                                                                                                                                                                                                                                                
Out[3]: True
In [4]: t.is_contiguous(memory_format=torch.channels_last)                                                                                                                                                                                                                                                                                                                                                                               
Out[4]: True
In [5]: t.is_contiguous(memory_format=torch.contiguous_format)                                                                                                                                                                                                                                                                                                                                                                           
Out[5]: True
```

**Comment by fhamborg** [comment]
plus one, getting the same warning when using huggingface's transformers:
``` 
...anaconda3\envs\newstsc2\lib\site-packages\transformers\modeling_xlnet.py:283: UserWarning: Mixed memory format inputs detected while calling the operator. The operator will output contiguous tensor even if some of the inputs are in channels_last format. (Triggered internally at  ..\aten\src\ATen\native\TensorIterator.cpp:918.)
  attn_score = (ac + bd + ef) * self.scale
```

**Comment by mranzinger** [comment]
It seems like have num_channels=1 is a surefire way to get this warning triggered.

**Comment by LuckyDC** [comment]
+1， same warning when conduct point-wise operation between 2 4-d tensors

**Comment by Rachnas** [comment]
> plus one, getting the same warning when using huggingface's transformers:
> 
> ```
> ...anaconda3\envs\newstsc2\lib\site-packages\transformers\modeling_xlnet.py:283: UserWarning: Mixed memory format inputs detected while calling the operator. The operator will output contiguous tensor even if some of the inputs are in channels_last format. (Triggered internally at  ..\aten\src\ATen\native\TensorIterator.cpp:918.)
>   attn_score = (ac + bd + ef) * self.scale
> ```
> 
> compare to [huggingface/transformers#7120](https://github.com/huggingface/transformers/issues/7120)

@fhamborg , I am also getting the same problem when trying to train xlnet for MNLI task. Do you have any update on this.
Thanks

**Comment by ad12** [comment]
Also reproducing this error when loading `state_dict` into a model (`model2`) when it is not on the same device as the model the `state_dict` originated from (`model1`).

**Script**:
```python
model1 = build_model()  # on cpu
model1 = model1.cuda()  # moved to cuda device

ckpt_path = "./ckpt1.pt"
torch.save(model1.state_dict(), ckpt_path)  # model1 params on cuda 

model2 = build_model()  # on cpu
model2.load_state_dict(torch.load(ckpt_path))  # loading params original on cuda onto cpu model
model2 = model2.cuda()
train(model2)  # some function that does training
```

```bash
# Warning
UserWarning: Mixed memory format inputs detected while calling the operator. The operator will output contiguous tensor even if some of the inputs are 
in channels_last format. (Triggered internally at  /opt/conda/conda-bld/pytorch_1595629427478/work/aten/src/ATen/native/TensorIterator.cpp:918.)
  output = self.final_layer(output) + input
```

This may be the case with huggingface transformer?

**Workaround**:
```python
model1 = build_model()  # on cpu
model1 = model1.cuda()  # moved to cuda device

ckpt_path = "./ckpt1.pt"
torch.save(model1.state_dict(), ckpt_path)  # model1 params on cuda 

model2 = build_model()  # on cpu
model2 = model2.cuda()  # move to cuda before loading state dict
model2.load_state_dict(torch.load(ckpt_path))
...
```

**Comment by JuriAbramov** [comment]
Guess this is new in PyTorch 1.7,  have a similar issue while PyTorch 1.6 works fine.

**Comment by alirezakazemipour** [comment]
> It is happening because ` tensor = box.reshape(N, -1, 4, H, W).permute(0, 3, 4, 1, 2)` outputs channels last 3d tensor. And when it comes in the mix with non channels last model this warning appears.
> 
> If this warning makes you nervous, you can add `tensor = box.reshape(N, -1, 4, H, W).permute(0, 3, 4, 1, 2).contiguous()` to make this warning disappear.

Thanks.
worked for me!

**Comment by sparseinference** [comment]
After switching to python3.8 I now get this warning.
This was very surprising after using python3.7 for a long time -- exactly the same code.

**Comment by hadaev8** [comment]
How to track what tensors exact have case this warning?

**Comment by sparseinference** [comment]
@alirezakazemipour I did try it with `.contiguous` after `.permute(...),` I still got the warning.

**Comment by alirezakazemipour** [comment]
@sparseinference
I did not use torchvision, so I have no idea about its relevancy to the issue. For me `.contiguous()` made the warning disappear.
Could you put the `permute()` part of your code here?

**Comment by sparseinference** [comment]
@alirezakazemipour It's very simple:

```python
def forward(self, u):
    u = u.permute(0,2,3,1).matmul(self.W0)
    u = u.permute(0,2,3,1).matmul(self.W1)
    u = u.permute(0,2,3,1).matmul(self.W2)
    return u
```
I just apply some matmuls on each dimension after permuting.
If I insert `.contiguous()` after every permute, I still get the warning. 
The input tensor shape is (B,C,H,W) , channels first.

**Comment by sparseinference** [comment]
This looks like the location in the code where the warning is coming from:

[https://github.com/pytorch/pytorch/blob/3e957d0545e650f45e11c75c4aac2313f200ff8b/aten/src/ATen/native/TensorIterator.cpp#L921](url)

**Comment by alirezakazemipour** [comment]
@sparseinference
I have no other suggestions but based on your input shape **(B, C, H, W)**, your permutation should be `permute(0, 3, 2, 1)` if the channel is at the third dimension, otherwise, there were no need to permute the input in your first line.
If you ever managed to fix it please update reponses here.
Good luck. :+1:
